From b7ff3a62494717a887a46dd3e304d30a2b2fa427 Mon Sep 17 00:00:00 2001
From: Jason Francis <cycl0ps@tuta.io>
Date: Sun, 18 Aug 2019 20:35:03 -0400
Subject: [PATCH] wayland: Error out instead of looping to find globals

Instead of spinning forever, do one roundtrip, which guarantees that the
global messages have been sent by the time we read the sync message.
If the proper globals aren't initialized yet, error out immediately.
---
 clutter/wayland/clutter-backend-wayland.c | 21 +++++++++++++++++----
 1 file changed, 17 insertions(+), 4 deletions(-)

diff --git a/clutter/wayland/clutter-backend-wayland.c b/clutter/wayland/clutter-backend-wayland.c
index 281b1a905..c1b7a6b10 100644
--- a/clutter/wayland/clutter-backend-wayland.c
+++ b/clutter/wayland/clutter-backend-wayland.c
@@ -215,10 +215,23 @@ clutter_backend_wayland_post_parse (ClutterBackend  *backend,
                             &wayland_registry_listener,
                             backend_wayland);
 
-  /* Wait until we have been notified about the compositor and shell objects */
-  while (!(backend_wayland->wayland_compositor &&
-           backend_wayland->wayland_shell))
-    wl_display_roundtrip (backend_wayland->wayland_display);
+  wl_display_roundtrip (backend_wayland->wayland_display);
+
+  if (!backend_wayland->wayland_compositor)
+    {
+      g_set_error (error, CLUTTER_INIT_ERROR,
+                  CLUTTER_INIT_ERROR_BACKEND,
+                  "Wayland compositor does not support wl_compositor");
+      return FALSE;
+    }
+
+  if (!backend_wayland->wayland_shell)
+    {
+      g_set_error (error, CLUTTER_INIT_ERROR,
+                  CLUTTER_INIT_ERROR_BACKEND,
+                  "Wayland compositor does not support wl_shell");
+      return FALSE;
+    }
 
   return TRUE;
 }
-- 
2.24.1

